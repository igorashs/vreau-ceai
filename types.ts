import FormidableLib from 'formidable';
import { ValidationError } from 'joi';

export type Product = {
  _id: string;
  name: string;
  category_id: string;
  src: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
};

export type ProductWithCategory = Product & {
  category_id: { name: string };
};

export type ErrorDetail = {
  message: string;
  name?: string;
};

export type ProductName = {
  name: string;
};

export type ProductNameErrorDetail = ErrorDetail & {
  name?: keyof ProductName;
};

export type ProductFields = ProductName & {
  src?: string;
  price: number;
  quantity: number;
  total_quantity: number;
  description: string;
  recommend: boolean;
  category_id: string;
};

export type Category = {
  _id: string;
  name: string;
};

export type ProductErrorDetail = ErrorDetail & {
  name?: keyof ProductFields;
};

export type ApiResponse = {
  success: boolean;
  message: string;
  errors?: ErrorDetail[];
};

export type User = {
  _id: string;
  name: string;
  email: string;
  isManager: boolean;
};

export type UserEmail = {
  email: string;
};

export type UserEmailErrorDetail = ErrorDetail & {
  name?: keyof UserEmail;
};

export type UserLogin = UserEmail & { password: string };

export type UserSignup = UserEmail & {
  name: string;
  password: string;
  repeat_password: string;
};

export type UserPermissions = { isManager: boolean };

export type LabelMessage = { success: boolean; message: string };

export type UserAuth = {
  _id: string;
  name: string;
  isAdmin: boolean;
  isManager: boolean;
};

export type UserSession = {
  isAuth: boolean;
  user: UserAuth | null;
  needRefresh: boolean;
};

type dataType = 'data' | 'error' | 'field' | 'fileBegin' | 'file' | 'progress';

export declare class Formidable extends FormidableLib {
  emit(name: dataType, data: FormidableLib.EmitData | ValidationError): void;
}

export type CategoryName = {
  name: string;
};

export type CategoryNameErrorDetail = ErrorDetail & {
  name?: keyof CategoryName;
};

export type OrderSubmit = {
  tel: string;
  address: string;
};

export type OrderItem = {
  product_id: string;
  count: number;
};

export type OrderStatus = {
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
};

export type OrderStatusErrorDetail = ErrorDetail & {
  name?: keyof OrderStatus;
};

export type OrderNumber = {
  number: string;
};

export type OrderNumberErrorDetail = ErrorDetail & {
  name?: keyof OrderNumber;
};

export type OrderFields = {
  info: OrderSubmit;
  items: OrderItem[];
};

export type Order = {
  _id: string;
  user: {
    name: string;
    email: string;
  };
  number: string;
  total_price: number;
  status: 'processing' | 'inDelivery' | 'canceled' | 'completed';
  items: {
    count: number;
    product: { name: string; price: number; quantity: number };
  }[];
  address: string;
  tel: string;
  orderedAt: Date;
  completedAt: Date;
};

export type UserOrders = { orders: Order[]; count: number };

export type ManagementOrders = { orders: Order[]; count: number };
