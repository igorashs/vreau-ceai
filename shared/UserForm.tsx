import styled from 'styled-components';
import Button from '@/shared/Button';
import { useForm } from 'react-hook-form';
import Form, { FormAction } from '@/shared/Form';
import CheckBox from '@/shared/CheckBox';
import { useEffect } from 'react';
import Label from '@/shared/Label';
import { User, UserPermissions } from 'types';

const UserData = styled.div`
  display: flex;
  flex-direction: column;
  gap: 7px;
  padding: 7px 0;
  border-bottom: 1px solid var(--layout);
`;

type UserFormProps = {
  onUserSubmit: (data: UserPermissions) => void;
  user: User;
};

const UserForm = ({ onUserSubmit, user }: UserFormProps) => {
  const { register, handleSubmit, reset } = useForm<UserPermissions>({
    mode: 'onChange',
    defaultValues: {
      isManager: user.isManager,
    },
  });

  useEffect(() => {
    reset({
      isManager: user.isManager,
    });
  }, [user]);

  return (
    <Form onSubmit={handleSubmit((data) => onUserSubmit(data))}>
      <div>
        <UserData>
          <Label>name</Label>
          {user.name}
        </UserData>
        <UserData>
          <Label>email</Label>
          {user.email}
        </UserData>
        <UserData>
          <Label>permisiuni</Label>
          <CheckBox label="manager" passRef={register} name="isManager" />
        </UserData>
      </div>
      <FormAction justify="flex-end">
        <Button>salveazÄƒ</Button>
      </FormAction>
    </Form>
  );
};

export default UserForm;
