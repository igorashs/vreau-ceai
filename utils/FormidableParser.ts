import * as validator from '@/utils/validator';
import FormidableNameSpace, { Files, File, IncomingForm } from 'formidable';
import { NextApiRequest } from 'next';
import { Formidable, ProductFields } from 'types';
import { productMessages } from './validator/schemas/product';
import { NextApiSessionRequest } from './withSession';

type ParsedData = {
  fields: Partial<ProductFields>;
  files: Files & { src?: File };
};

type HandleProductForm = (
  req: NextApiRequest | NextApiSessionRequest,
  options: Partial<FormidableNameSpace.Options>,
  allowedFileTypes: string[],
) => Promise<ParsedData>;

const handleProductForm: HandleProductForm = async (
  req,
  options,
  allowedFileTypes,
) => {
  const data: ParsedData = await new Promise((resolve, reject) => {
    const form: Formidable = new IncomingForm(options);

    form.onPart = (part) => {
      if (part.filename) {
        // get file extension
        const fileType = part.mime?.split('/').pop() ?? '';

        // validate allowed file types
        if (!allowedFileTypes.includes(fileType)) {
          form.emit(
            'error',
            validator.createValidationError({
              message: productMessages.src.invalid,
              key: 'src',
            }),
          );
        }
      }

      form.handlePart(part);
    };

    form.parse(req, (err, fields, files) => {
      if (err) {
        // custom error on maxFileSize
        if (err.message.includes('maxFileSize')) {
          reject(
            validator.createValidationError({
              message: productMessages.src.max,
              key: 'src',
            }),
          );
        } else {
          reject(err);
        }
      }

      resolve({ fields, files });
    });
  });

  return data;
};

export default {
  handleProductForm,
};
