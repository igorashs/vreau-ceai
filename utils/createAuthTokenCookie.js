import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createAuthTokenCookie = (payload) => {
  const authToken = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  return cookie.serialize('auth-token', authToken, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });
};
