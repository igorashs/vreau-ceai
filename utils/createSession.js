import jwt from 'jsonwebtoken';
import cookie from 'cookie';

const EXPIRE_TIME = 24 * 60 * 60;

export const createSession = (payload) => {
  const token = jwt.sign(payload, process.env.AUTH_PRIVATE_KEY, {
    expiresIn: EXPIRE_TIME
  });

  const [header, jwtPayload, signature] = token.split('.');
  const session = header + '.' + jwtPayload;

  const sessionCookie = cookie.serialize('session', session, {
    secure: process.env.NODE_ENV !== 'development',
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  const authCookie = cookie.serialize('auth', signature, {
    secure: process.env.NODE_ENV !== 'development',
    httpOnly: true,
    sameSite: 'strict',
    path: '/',
    maxAge: EXPIRE_TIME
  });

  return [sessionCookie, authCookie];
};
