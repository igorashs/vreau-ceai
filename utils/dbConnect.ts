import mongoose from 'mongoose';

const dbConnect = async () => {
  if (mongoose.connection.readyState >= 1) {
    return;
  }

  await mongoose.connect(process.env.MONGODB_URI as string, {
    useNewUrlParser: true,
    useUnifiedTopology: true,
    useCreateIndex: true,
    useFindAndModify: false,
  });
};

export default dbConnect;
