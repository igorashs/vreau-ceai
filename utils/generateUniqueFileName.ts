import { nanoid } from 'nanoid';

const generateUniqueFileName = (fileName: string) => {
  // grab file extension
  const matches = fileName?.match(/\..+$/);

  // create new file name
  const uniqueFileName = nanoid() + (matches && matches[0]);

  return uniqueFileName;
};

export default generateUniqueFileName;
