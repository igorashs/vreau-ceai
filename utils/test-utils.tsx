import { render, RenderOptions } from '@testing-library/react';
import GlobalStyle from 'GlobalStyle';

const Wrapper: React.FC = ({ children }) => (
  <>
    <GlobalStyle />
    {children}
  </>
);

const customRender = (
  ui: React.ReactElement,
  options?: Omit<RenderOptions, 'queries'>,
) => render(ui, { wrapper: Wrapper, ...options });

export * from '@testing-library/react';

export { customRender as render };
