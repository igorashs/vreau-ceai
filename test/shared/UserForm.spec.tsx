import { fireEvent, render, waitFor } from '@/utils/test-utils';
import UserForm from '@/shared/UserForm';
import { User } from 'types';

const onUserSubmitMock = jest.fn();

describe('UserForm', () => {
  it('renders correctly', () => {
    const user: User = {
      _id: '1',
      email: 'user@user.com',
      isManager: true,
      name: 'user',
    };

    const { container, getByText, getByRole } = render(
      <UserForm onUserSubmit={onUserSubmitMock} user={user} />,
    );

    expect(getByText(user.email)).toBeVisible();
    expect(getByText(user.name)).toBeVisible();
    expect(getByRole('checkbox')).toBeChecked();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('success submission', () => {
    it('calls with expected values', async () => {
      const user: User = {
        _id: '1',
        email: 'user@user.com',
        isManager: true,
        name: 'user',
      };

      const { getByRole } = render(
        <UserForm onUserSubmit={onUserSubmitMock} user={user} />,
      );

      fireEvent.click(getByRole('checkbox'));
      fireEvent.submit(getByRole('button'));

      await waitFor(() =>
        expect(onUserSubmitMock).toBeCalledWith({ isManager: false }),
      );

      expect(getByRole('checkbox')).not.toBeChecked();
    });
  });

  // ? currently doesn't have any validation error
});
