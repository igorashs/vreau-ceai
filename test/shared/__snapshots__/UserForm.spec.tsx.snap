// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserForm renders correctly 1`] = `
.c7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  gap: 7px;
  padding: calc(var(--baseline) / 4) calc(var(--baseline) / 2);
  border-radius: 4px;
  border: 0;
  color: var(--text-light);
  background-color: var(--accent-dark);
  box-shadow: 1px 1px 2px #00000066;
  text-transform: lowercase;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.c7:hover {
  background-color: var(--accent);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: calc(var(--baseline) / 2);
}

.c6 {
  margin-top: calc(var(--baseline) / 2);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -webkit-justify-content: flex-end;
  -ms-flex-pack: end;
  justify-content: flex-end;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c2 {
  font-weight: 500;
}

.c4 {
  font-weight: 500;
  color: var(--accent-text-dark);
  font-weight: 400;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: 7px;
}

.c5 {
  margin: 0;
  width: 16px;
  height: 16px;
  -webkit-order: -1;
  -ms-flex-order: -1;
  order: -1;
  cursor: pointer;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: 7px;
  padding: 7px 0;
  border-bottom: 1px solid var(--layout);
}

<form
  class="c0"
>
  <div>
    <div
      class="c1"
    >
      <label
        class="c2"
      >
        name
      </label>
      user
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
      >
        email
      </label>
      user@user.com
    </div>
    <div
      class="c1"
    >
      <label
        class="c2"
      >
        permisiuni
      </label>
      <div
        class="c3"
      >
        <label
          class="c4"
          for="isManager"
        >
          manager
        </label>
        <input
          class="c5"
          id="isManager"
          name="isManager"
          type="checkbox"
        />
      </div>
    </div>
  </div>
  <div
    class="c6"
  >
    <button
      class="c7"
    >
      salveazÄƒ
    </button>
  </div>
</form>
`;
